libs/cosmos-sdk/baseapp/abci.go

```diff
@@ -391,7 +392,15 @@ func handleQueryApp(app *BaseApp, path []string, req abci.RequestQuery) abci.Res
 	if len(path) >= 2 {
 		switch path[1] {
 		case "simulate":
-			txBytes := req.Data
+			queryBytes := req.Data
+			var queryData types.SimulateData
+			var txBytes []byte
+			if err := json.Unmarshal(queryBytes, &queryData); err != nil {
```

Comment:
Use a new case here? Such as simulateWithState, and only decode json in the new case for better performance.